src/Microsoft.IdentityModel.Tokens/SymmetricSignatureProvider.cs

```diff
@@ -238,20 +226,23 @@ public override bool Verify(byte[] input, byte[] signature)
                 throw LogHelper.LogExceptionMessage(new ObjectDisposedException(GetType().ToString()));
             }
 
-
             LogHelper.LogInformation(LogMessages.IDX10643, input);
+            KeyedHashAlgorithm keyedHashAlgorithm = GetKeyedHashAlgorithm(GetKeyBytes(Key), Algorithm);
             try
             {
-                lock (_verifyLock)
-                {
-                    return Utility.AreEqual(signature, KeyedHashAlgorithm.ComputeHash(input));
-                }
+                return Utility.AreEqual(signature, keyedHashAlgorithm.ComputeHash(input));
             }
             catch
             {
                 CryptoProviderCache?.TryRemove(this);
+                Dispose(true);
```

Comment:
Suppose two threads are using the object pool

T1 acquires an element from the pool
T2 acquires an element and throws and hits this, disposing the the element that T1 acquires, T1 throws an object disposed exception on use.

Applies more so with larger object pools.

Am I missing something here? Is there anything that we can do in that case?

Reply:
@keegan-caruso T1 and T2 will have different instances of the element, that is the idea of the object pool.
As George pointed out below, we will have to think about when an RSA or ECD object is passed to the constructor of the SecurityKey, we may have to recreate a new object. However that could be cause issues with derived types.